lib/Sema/AvailableFactsAnalysis.cpp

```diff
@@ -0,0 +1,499 @@
+ if (S2.size() == 0)
+    return S1;
+  ComparisonSet Result;
+  for (auto E1 : S1)
+    if (S2.find(E1) == S2.end())
+      Result.insert(E1);
+  return Result;
+}
+
+// Given two sets S1 and S2, the return value is the union of these sets.
+ComparisonSet AvailableFactsAnalysis::Union(ComparisonSet &S1, ComparisonSet &S2) {
+  if (S1.size() == 0)
+    return S2;
+  if (S2.size() == 0)
+    return S1;
+  ComparisonSet Result(S1);
+  Result.insert(S2.begin(), S2.end());
+  return Result;
+}
+
+// Given two sets S1 and S2, this function returns true if the two sets
+// are not equal.
+// Equality means that their sizes are the same, and every member of S1 is found in
+// S2 and vice versa.
+bool AvailableFactsAnalysis::Differ(ComparisonSet &S1, ComparisonSet &S2) {
+  if (S1.size() != S2.size())
+    return true;
+  for (auto E : S1)
+    if (S2.find(E) == S2.end())
+      return true;
+  return false;
+}
+
+// Given two sets S1 and S2, the return value is the intersection of these sets.
+ComparisonSet AvailableFactsAnalysis::Intersect(ComparisonSet &S1, ComparisonSet &S2) {
+  if (S1.size() == 0)
+    return S1;
+  if (S2.size() == 0)
+    return S2;
+  ComparisonSet Result;
+  for (auto E1 : S1)
+    if (S2.find(E1) != S2.end())
+      Result.insert(E1);
+  return Result;
+}
+
+// This function returns true if variable `V` is used in the comparison `I`.
+bool AvailableFactsAnalysis::ContainsVariable(Comparison& I, const VarDecl *V) {
+  std::set<const Expr *> Exprs;
+  CollectExpressions(I.first, Exprs);
+  CollectExpressions(I.second, Exprs);
+  for (auto InnerExpr : Exprs)
+    if (const DeclRefExpr *DRE = dyn_cast<DeclRefExpr>(InnerExpr))
+      if (const VarDecl *VD = dyn_cast<VarDecl>(DRE->getDecl()))
+        if (VD == V)
+          return true;
+  return false;
+}
+
+// This function returns true only if expression `E` is a pointer deref.
+// A pointer deref is defined as follows:
+// - `E` should be a cast of type `LValueToRValue, and
+// - If the child of `E` is `e`, then `f(e)` must be true. `f(e)` is defined in
+//   a separate function `IsPointerDerefLValue`.
+// Otherwise, the function returns false;
+bool AvailableFactsAnalysis::IsPointerDeref(const Expr *E) {
+  if (const CastExpr *CE = dyn_cast<CastExpr>(E)) {
+    if (CE->getCastKind() != CastKind::CK_LValueToRValue)
+      return false;
+    return IsPointerDerefLValue(CE->getSubExpr());
+  }
+  return false;
+}
+
+bool AvailableFactsAnalysis::IsPointerAssignment(const Expr *E) {
+  if (const BinaryOperator *BO = dyn_cast<BinaryOperator>(E)) {
+    if (BO->isAssignmentOp())
+      if (IsPointerDerefLValue(BO->getLHS()))
+        return true;
+  }
+  if (const UnaryOperator *UO = dyn_cast<UnaryOperator>(E)) {
+    if (UO->isIncrementDecrementOp())
+      if (IsPointerDerefLValue(UO->getSubExpr()))
+        return true;
+  }
+  if (const CallExpr *CE = dyn_cast<CallExpr>(E)) {
+    for (int ArgIndex = 0; ArgIndex < CE->getNumArgs(); ArgIndex++) {
```

Comment:
Any call expression must be considered as possibly doing a pointer assignment.   We should not consider the types of the arguments passed to the function call.  It's possible that a pointer could be accessible to a function via a global variable or through a structure type.

Reply:
Fixed.